File: project/jeonga-onboarding/src/test/java/onboarding/survey/service/SurveyServiceTest.java
package onboarding.survey.service;

import jakarta.transaction.Transactional;
import onboarding.survey.domain.InputType;
import onboarding.survey.dto.CreateSurveyRequest;
import onboarding.survey.dto.SurveyItemDto;
import onboarding.survey.exception.BadRequestException;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;
import java.util.stream.IntStream;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

@SpringBootTest
@Transactional
class SurveyServiceTest {

    @Autowired
    SurveyService surveyService;

    @Test
    void 정상_생성_테스트() {
        var itemDto = new SurveyItemDto(
                "Q1", "설명", InputType.SINGLE_CHOICE, true, List.of("A", "B")
        );
        var request = new CreateSurveyRequest("테스트", "설명", List.of(itemDto));
        Long id = surveyService.createSurvey(request);
        assertNotNull(id);
    }

    @Test
    void 항목_0개_생성시_예외() {
        var request = new CreateSurveyRequest("테스트", "설명", List.of());
        assertThrows(BadRequestException.class, () -> surveyService.createSurvey(request));
    }

    @Test
    void 선택형_선택지_없으면_예외() {
        var itemDto = new SurveyItemDto(
                "Q1", "설명", InputType.SINGLE_CHOICE, true, List.of()
        );
        var request = new CreateSurveyRequest("테스트", "설명", List.of(itemDto));
        assertThrows(BadRequestException.class, () -> surveyService.createSurvey(request));
    }

    @Test
    void 항목_10개_초과시_예외() {
        var itemDtos = IntStream.range(0, 11)
                .mapToObj(i -> new SurveyItemDto("Q" + i, "desc", InputType.SHORT_TEXT, true, null))
                .toList();
        var request = new CreateSurveyRequest("테스트", "설명", itemDtos);
        assertThrows(BadRequestException.class, () -> surveyService.createSurvey(request));
    }
}

File: project/jihochoi-onboarding/QueryBox/src/test/kotlin/com/wlghsp/querybox/application/SurveyServiceTest.kt
package com.wlghsp.querybox.application

import com.wlghsp.querybox.domain.survey.QuestionType
import com.wlghsp.querybox.ui.dto.QuestionRequest
import com.wlghsp.querybox.ui.dto.QuestionUpdateRequest
import com.wlghsp.querybox.ui.dto.SurveyCreateRequest
import com.wlghsp.querybox.ui.dto.SurveyUpdateRequest
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import

@DataJpaTest
@Import(SurveyService::class)
class SurveyServiceTest @Autowired constructor(
    val surveyService: SurveyService,
) {

    @DisplayName("설문 생성 성공")
    @Test
    fun createSurvey_success() {
        val request = SurveyCreateRequest(
            title = "테스트 설문",
            description = "설명입니다",
            questions = listOf(
                QuestionRequest(
                    name = "언어?",
                    description = "좋아하는 언어는?",
                    type = QuestionType.SHORT_TEXT,
                    required = true,
                    options = null
                )
            )
        )

        val id = surveyService.create(request)

        val survey = surveyService.findSurveyWithQuestionsById(id)
        assertThat(survey.title).isEqualTo("테스트 설문")
        assertThat(survey.getQuestions()).hasSize(1)
    }

    @DisplayName("설문 수정 성공")
    @Test
    fun updateSurvey_success() {
        // given
        val id = surveyService.create(
            SurveyCreateRequest(
                title = "원래 제목",
                description = "원래 설명",
                questions = listOf(
                    QuestionRequest(
                        name = "언어?",
                        description = "언어 질문",
                        type = QuestionType.SHORT_TEXT,
                        required = true,
                        options = null
                    )
                )
            )
        )

        val survey = surveyService.findSurveyWithQuestionsById(id)
        val questionId = survey.getQuestions().first().id

        // when
        val updateRequest = SurveyUpdateRequest(
            title = "변경된 제목",
            description = "변경된 설명",
            questions = listOf(
                QuestionUpdateRequest(
                    id = questionId,
                    name = "바뀐 질문",
                    description = "바뀐 설명",
                    type = QuestionType.SHORT_TEXT,
                    required = false,
                    options = null
                )
            )
        )
        surveyService.update(id, updateRequest)


        val updated = surveyService.findSurveyWithQuestionsById(id)
        assertThat(updated.title).isEqualTo("변경된 제목")
        assertThat(updated.getQuestions().first().name).isEqualTo("바뀐 질문")
    }

    @DisplayName("설문 id로 설문 조회 가능")
    @Test
    fun findSurveyWithQuestionsById_success() {
        val id = surveyService.create(
            SurveyCreateRequest(
                title = "조회 테스트",
                description = "조회 설명",
                questions = listOf(
                    QuestionRequest(
                        name = "이름은?",
                        description = "감자",
                        type = QuestionType.SHORT_TEXT,
                        required = false,
                        options = null
                    )
                ),
            )
        )

        val found = surveyService.findSurveyWithQuestionsById(id)

        assertThat(found.title).isEqualTo("조회 테스트")
        assertThat(found.description).isEqualTo("조회 설명")
    }
}